lighthouse-core/scripts/i18n/bake-strings.js

```diff
@@ -0,0 +1,144 @@
+ */
+
+/**
+ * @typedef ICUPlaceholderDefn
+ * @property {string} content
+ * @property {string} [example]
+ */
+
+const ignoredPathComponents = [
+  '/.git',
+  '/scripts',
+  '/node_modules',
+  '/test/',
+  '-test.js',
+  '-renderer.js',
+];
+
+/**
+ * Take a series of CTC format ICU messages and converts them to LHL format by
+ * replacing $placeholders$ with their {ICU} values. Functional opposite of
+ * `convertMessageToPlaceholders`. This is commonly called as the last step in
+ * translation.
+ *
+ * Converts this:
+ * messages: {
+ *  "lighthouse-core/audits/seo/canonical.js | explanationDifferentDomain" {
+ *    "message": "Points to a different domain ($ICU_0$)",
+ *    "placeholders": {
+ *      "ICU_0": {
+ *        "content": "{url}",
+ *        "example": "https://example.com/"
+ *      },
+ *    },
+ *  },
+ * }
+ *
+ * Into this:
+ * messages: {
+ *  "lighthouse-core/audits/seo/canonical.js | explanationDifferentDomain" {
+ *    "message": "Points to a different domain ({url})",
+ *    },
+ *  },
+ * }
+ *
+ * Throws if there is a $placeholder$ in the message that has no corresponding
+ * value in the placeholders object, or vice versa.
+ *
+ * @param {Record<string, ICUMessageDefn>} messages
+ * @returns {Record<string, ICUMessageDefn>}
+ */
+function bakePlaceholders(messages) {
+  for (const [_, defn] of Object.entries(messages)) {
+    // Don't need descriptions or meanings anymore.
+    delete defn['description'];
+    delete defn['meaning'];
+
+    let message = defn['message'];
+    const placeholders = defn['placeholders'];
+
+    for (const placeholder in placeholders) {
+      if (!Object.prototype.hasOwnProperty.call(placeholders, placeholder)) continue;
+
+      const content = placeholders[placeholder]['content'];
+      if (!message.includes(`$${placeholder}$`)) {
+        throw Error(`Message "${message}" has extra placeholder "${placeholder}"`);
+      }
+      message = message.replace(`$${placeholder}$`, content);
+    }
+
+    // Sanity check that all placeholders are gone
+    if (message.match(/\$\w+\$/)) throw Error(`Message "${message}" is missing placeholder`);
+
+    defn['message'] = message;
+
+    delete defn['placeholders'];
+  }
+  return messages;
+}
+
+/**
+ * @param {*} file
+ */
+function collectPreLocaleStrings(file) {
+  const rawdata = fs.readFileSync(file, 'utf8');
+  const messages = JSON.parse(rawdata);
+  return messages;
+}
+
+/**
+ * @param {*} path
+ * @param {*} output
+ */
+function saveLocaleStrings(path, output) {
+  fs.writeFileSync(path, JSON.stringify(output, null, 2) + '\n');
+}
+
+/**
+ * @param {string} dir
+ * @param {string} output
+ * @returns {Array<string>}
+ */
+function collectAndBakeCtcStrings(dir, output) {
+  const lhl = [];
+  for (const name of fs.readdirSync(dir)) {
+    const fullPath = path.join(dir, name);
+    const relativePath = path.relative(LH_ROOT, fullPath);
+    if (ignoredPathComponents.some(p => fullPath.includes(p))) continue;
```

Comment:
as mentioned above, this check seems unnecessary due to the `'.ctc.json'` check below